---
title: "Comparison: Michigan Imputation Server / Shapeit+Minimac3 (Chromosome 7)"
author: "Ignacio Tolosana"
output: pdf_document
---

# Imputation

```{r echo=FALSE}
library(VariantAnnotation)
library(ggplot2)
library(snpStats)
library(scoreInvHap)
library(BiocParallel)
library(ggpubr)
source("InversionNGSutils.R")

load(file = "michigan_chr7.rda")
load(file = "michigan_invclust_chr7.rda")
load(file = "michigan_scoreinvhap_chr7.rda")
load(file = "minimac_chr7.rda")
load(file = "minimac_invclust_chr7.rda")
load(file = "minimac_scoreinvhap_chr7.rda")
load(file = "nofilter_michigan_7.rda")
load(file = "nofilter_michigan_inv_7.rda")
load(file = "nofilter_michigan_invclust_7.rda")
load(file = "nofilter_minimac_7.rda")
load(file = "nofilter_minimac_inv_7.rda")
load(file = "nofilter_minimac_invclust_7.rda")
```

## Imputed data exploration
```{r}
michigan_chr7
minimac_chr7
```


## DS values
```{r echo=FALSE}
DS_michigan <- geno(michigan_chr7)$DS
DS_minimac <- geno(minimac_chr7)$DS
```

```{r}
# Distribution of the DS values in each imputation
par(mfrow=c(1,2))
HIST_MICHIGAN <- hist(DS_michigan, breaks=seq(0, 2, by=0.05), 
                      main="Michigan Imputation Server", xlab="DS values")
HIST_MINIMAC <- hist(DS_minimac, breaks=seq(0, 2, by=0.05), 
                     main="Minimac3", xlab="DS values")
```


```{r echo=FALSE}
#Get SNPs and individuals present in both methods
DS_rows <- intersect(rownames(DS_michigan), rownames(DS_minimac))
DS_col <- intersect(colnames(DS_michigan), colnames(DS_minimac))

DS_michigan_common <- DS_michigan[DS_rows,DS_col]
DS_minimac_common <- DS_minimac[DS_rows,DS_col]

#Calculate DS correlation by individuals
cor_by_ind <- mapply(cor, as.data.frame(DS_michigan_common),
                          as.data.frame(DS_minimac_common))

```

```{r}
# DS correlation by individuals
min(cor_by_ind)
max(cor_by_ind)
mean(cor_by_ind)
mean(cor_by_ind > 0.95)
mean(cor_by_ind > 0.99)
```


```{r}
# Histogram of the DS correlation values by individuals
par(mfrow=c(1,1))
CORR_HIST <- hist(cor_by_ind, breaks=seq(0.48, 1, by=0.01), 
                  main="DS correlation values by individuals", xlab="DS correlation")
```



## R^2^
```{r echo=FALSE}
rsq <- c(info(michigan_chr7)$R2, info(minimac_chr7)$R2)
ismich <- rep("Michigan", length(info(michigan_chr7)$R2))
ismich <- c(ismich, rep("Minimac", length(info(minimac_chr7)$R2)))
comparison <- data.frame(rsq, ismich)
```


Density and histogram plots comparing the RS2 values in both methods ("ismich = TRUE" indicates the values for the Michigan imputation, whereas "ismich = FALSE" shows the values for the Minimac imputation)
```{r}
ggdensity(comparison, x = "rsq",
          add = "mean", rug = TRUE,
          color = "ismich", fill = "ismich",
          palette = c("#E7B800", "#00AFBB"),
          legend.title = c(""), 
          xlab = ("Imputation R2"),
          ylab = ("Density"))

par(mfrow=c(1,2))
HIST_R2_MICHIGAN <- hist(info(michigan_chr7)$R2, breaks=seq(0.5, 1, by=0.02), 
                         main="Michigan", xlab="Imputation R2")
HIST_R2_MINIMAC <- hist(info(michigan_chr7)$R2, breaks=seq(0.5, 1, by=0.02), 
                        main="Minimac", xlab="Imputation R2")
```


## Genotype predictions


```{r echo=FALSE, warning=FALSE}
#Convert genotypes into matrix
gen_michigan <- genotypeToSnpMatrix(michigan_chr7)
matrix_gen_michigan <- t(as(gen_michigan$genotype, "character"))

gen_minimac <- genotypeToSnpMatrix(minimac_chr7)
matrix_gen_minimac <- t(as(gen_minimac$genotype, "character"))

#Compare the genotype predictions with each method by individuals
common_rows <- intersect(rownames(matrix_gen_michigan), rownames(matrix_gen_minimac))
common_col <- intersect(colnames(matrix_gen_michigan), colnames(matrix_gen_minimac))

matrix_common_michigan <- matrix_gen_michigan[common_rows, common_col]
matrix_common_minimac <- matrix_gen_minimac[common_rows, common_col]

perc_by_ind <- c()

for (i in common_col){
  genotype_by_ind <- c(matrix_common_michigan[,i] == matrix_common_minimac[,i])
  perc_by_ind <- c(perc_by_ind, mean(genotype_by_ind))
}
```

Compare the genotype predictions (BestGuess) with each method by individuals. "perc_by_ind" is the % of SNPs by individual predicted equally in both methods
```{r}
min(perc_by_ind)
max(perc_by_ind)
mean(perc_by_ind)
mean(perc_by_ind > 0.95)
mean(perc_by_ind > 0.99)
```


```{r}
# Plot histogram of the genotypes equally predicted (BestGuess) by individuals 
# in both methods
par(mfrow=c(1,1))
GENO_HIST <- hist(perc_by_ind, breaks=seq(0.7, 1, by=0.01), 
                  main="SNPs (genotypes) equally predicted with Michigan and Minimac3", 
                  xlab="% of equally predicted genotypes")
```



# Inversion prediction

```{r echo=FALSE}
#Load reference data scoreInvHap
data("Refs")
ref <- Refs$inv7p11.2
data("SNPsR2")
R2s <- SNPsR2$inv7p11.2
data("hetRefs")
hRefs <- hetRefs$inv7p11.2
```

Predicted inversions with scoreInvHap
```{r}
michigan_inv_chr7
minimac_inv_chr7
```

```{r echo=FALSE}
scoreinvhap_table <- table(Michigan = classification(michigan_inv_chr7)[sort(names(classification(michigan_inv_chr7)))], 
      Minimac = classification(minimac_inv_chr7)[sort(names(classification(minimac_inv_chr7)))])
```

```{r}
# Comparison table
scoreinvhap_table
sum(diag(scoreinvhap_table))/sum(scoreinvhap_table)
```


```{r}
# Comparison of the results for both imputation methods
par(mfrow=c(1,2))
hist(maxscores(michigan_inv_chr7), breaks=seq(0, 1, by=0.05), main="Michigan", xlab="Maxscores")
hist(maxscores(minimac_inv_chr7), breaks=seq(0, 1, by=0.05), main="Minimac", xlab="Maxscores")
```

```{r echo=FALSE}
# Calculate scores correlation by individuals
score_corr <- mapply(cor, as.data.frame(scores(michigan_inv_chr7)[sort(rownames(scores(michigan_inv_chr7))),]),
                          as.data.frame(scores(minimac_inv_chr7)[sort(rownames(scores(minimac_inv_chr7))),]))
```

```{r}
# Score correlation by individuals between both imputation methods
min(score_corr)
max(score_corr)
mean(score_corr)
par(mfrow=c(1,1))
SCORE_CORR_HIST <- hist(score_corr, breaks=seq(0.9, 1, by=0.005), 
                        main="Score correlation by individuals", 
                        xlab="Score correlation")
```

```{r}
# Difference score between the highest similarity score and the second highest, 
# in both imputation methods
par(mfrow=c(1,2))
hist(diffscores(michigan_inv_chr7), breaks=seq(0, 1, by=0.05), 
     main="Michigan", xlab="Difference score")
hist(diffscores(minimac_inv_chr7), breaks=seq(0, 1, by=0.05), 
     main="Minimac", xlab="Difference score")
```


```{r}
# Numbers of scores used
mean(numSNPs(michigan_inv_chr7))
mean(numSNPs(minimac_inv_chr7))
```

```{r}
# Number of samples in both imputation methods before and after QC filtering
length(classification(michigan_inv_chr7))
length(classification(michigan_inv_chr7, minDiff = 0.1, callRate = 0.9))
length(classification(michigan_inv_chr7, minDiff = 0.1, callRate = 0.9))/
  length(classification(michigan_inv_chr7))
length(classification(minimac_inv_chr7))
length(classification(minimac_inv_chr7, minDiff = 0.1, callRate = 0.9))
length(classification(minimac_inv_chr7, minDiff = 0.1, callRate = 0.9))/
  length(classification(minimac_inv_chr7))
```



# Plots with invClust


```{r}
# Michigan
par(mfrow=c(1,1))
plotInv(michigan_invclust_chr7, classification = classification(michigan_inv_chr7))
```


```{r}
# Minimac
plotInv(minimac_invclust_chr7, classification = classification(minimac_inv_chr7))
```


# No filtered imputed data

```{r}
nofilter_minimac_7
nofilter_michigan_7
nofilter_michigan_inv_7
nofilter_minimac_inv_7

# Select SNPs in both elements to represent them in the plot
snps_minimac_7 <- intersect(rownames(info(nofilter_minimac_7)), names(SNPsR2$inv7p11.2))

# Plot Imputation R2 vs scoreInvHap R2 (red line = filter in the previous data)
ggplot() +
  geom_point(aes(x = info(nofilter_minimac_7)[snps_minimac_7,]$R2, 
                 y = SNPsR2$inv7p11.2[snps_minimac_7]), 
                alpha = 0.2) +
  geom_vline(aes(xintercept=0.5), colour="red") +
  ggtitle("Minimac chr 7") +
  xlab("Imputation R2") +
  ylab("scoreInvHap R2")

hist(classification(nofilter_minimac_inv_7), 
     main="No filtered chr7", xlab="Genotype")


# Histograms of scoreInvHap R2 in the filtered imputed data and in the NO filtered imputed data
par(mfrow=c(1,2))
hist(SNPsR2$inv7p11.2[rownames(minimac_chr7)], breaks=seq(0, 1, by=0.02), 
     main="Filtered chr7", xlab="scoreInvHapR2")
hist(SNPsR2$inv7p11.2[rownames(nofilter_minimac_7)], breaks=seq(0, 1, by=0.02), 
     main="No filtered chr7", xlab="scoreInvHapR2")

#Correlation between Imputation R2 and scoreInvHap R2 (NO filtered data)
cor(info(nofilter_minimac_7)[snps_minimac_7,]$R2, SNPsR2$inv7p11.2[snps_minimac_7])
```

```{r}
# invClust plot of the NO filtered imputed data (minimac)
plotInv(nofilter_minimac_invclust_7, 
        classification = classification(nofilter_minimac_inv_7))
# invClust plot of the NO filtered imputed data (michigan)
# (invClust only used 96 SNPs in this case)
plotInv(nofilter_michigan_invclust_7, 
        classification = classification(nofilter_michigan_inv_7))
```



