---
title: '"imputeinversion" results'
author: "Ignacio Tolosana"
subtitle: "Chromosome 17"
output: html_document
---

# Imputation

```{r echo=FALSE}
library(VariantAnnotation)
library(ggplot2)
library(snpStats)
library(scoreInvHap)
library(BiocParallel)
library(ggpubr)
source("InversionNGSutils.R")
```

## Imputed data exploration
```{r}
byregion
whole
```


## DS values
```{r echo=FALSE}
DS_byregion <- geno(byregion)$DS
DS_whole <- geno(whole)$DS
```

```{r}
# Distribution of the DS values in each imputation
par(mfrow=c(1,2))
HIST_BYREGION <- hist(DS_byregion, breaks=seq(0, 2, by=0.05), 
                      main="Imputation by region", xlab="DS values")
HIST_WHOLE <- hist(DS_whole, breaks=seq(0, 2, by=0.05), 
                     main="Imputation whole chr", xlab="DS values")
```


```{r echo=FALSE}
#Get SNPs and individuals present in both methods
DS_rows <- intersect(rownames(DS_byregion), rownames(DS_whole))
DS_col <- intersect(colnames(DS_byregion), colnames(DS_whole))

DS_byregion_common <- DS_byregion[DS_rows,DS_col]
DS_whole_common <- DS_whole[DS_rows,DS_col]

#Calculate DS correlation by individuals
cor_by_ind <- mapply(cor, as.data.frame(DS_byregion_common),
                          as.data.frame(DS_whole_common))

```

```{r}
# DS correlation by individuals
min(cor_by_ind)
max(cor_by_ind)
mean(cor_by_ind)
mean(cor_by_ind > 0.95)
mean(cor_by_ind > 0.99)
```


```{r}
# Histogram of the DS correlation values by individuals
par(mfrow=c(1,1))
CORR_HIST <- hist(cor_by_ind, breaks=seq(0.85, 1, by=0.002), 
                  main="DS correlation values by individuals", xlab="DS correlation")
```



## R^2^
```{r echo=FALSE}
rsq <- c(info(byregion)$R2, info(whole)$R2)
ismich <- rep("ByRegion", length(info(byregion)$R2))
ismich <- c(ismich, rep("WholeChr", length(info(whole)$R2)))
comparison <- data.frame(rsq, ismich)
```


Density and histogram plots comparing the RS2 values in both methods ("ismich = TRUE" indicates the values for the Michigan imputation, whereas "ismich = FALSE" shows the values for the Minimac imputation)
```{r}
ggdensity(comparison, x = "rsq",
          add = "mean", rug = TRUE,
          color = "ismich", fill = "ismich",
          palette = c("#E7B700", "#00AFBB"),
          legend.title = c(""), 
          xlab = ("Imputation R2"),
          ylab = ("Density"))

par(mfrow=c(1,2))
HIST_R2_BYREGION <- hist(info(byregion)$R2, breaks=seq(0, 1, by=0.05), 
                         main="ByRegion", xlab="Imputation R2")
HIST_R2_WHOLE <- hist(info(whole)$R2, breaks=seq(0, 1, by=0.05), 
                        main="Whole", xlab="Imputation R2")
```


## Genotype predictions


```{r echo=FALSE, warning=FALSE}
#Convert genotypes into matrix
gen_byregion <- genotypeToSnpMatrix(byregion)
matrix_gen_byregion <- t(as(gen_byregion$genotype, "character"))

gen_whole <- genotypeToSnpMatrix(whole)
matrix_gen_whole <- t(as(gen_whole$genotype, "character"))

#Compare the genotype predictions with each method by individuals
common_rows <- intersect(rownames(matrix_gen_byregion), rownames(matrix_gen_whole))
common_col <- intersect(colnames(matrix_gen_byregion), colnames(matrix_gen_whole))

matrix_common_byregion <- matrix_gen_byregion[common_rows, common_col]
matrix_common_whole <- matrix_gen_whole[common_rows, common_col]

perc_by_ind <- c()

for (i in common_col){
  genotype_by_ind <- c(matrix_gen_byregion[,i] == matrix_gen_whole[,i])
  perc_by_ind <- c(perc_by_ind, mean(genotype_by_ind))
}
```

Compare the genotype predictions (BestGuess) with each method by individuals. "perc_by_ind" is the % of SNPs by individual predicted equally in both methods
```{r}
min(perc_by_ind)
max(perc_by_ind)
mean(perc_by_ind)
mean(perc_by_ind > 0.95)
mean(perc_by_ind > 0.99)
```


```{r}
# Plot histogram of the genotypes equally predicted (BestGuess) by individuals 
# in both methods
par(mfrow=c(1,1))
GENO_HIST <- hist(perc_by_ind, breaks=seq(0.9, 1, by=0.002), 
                  main="SNPs (genotypes) equally predicted", 
                  xlab="% of equally predicted genotypes")
```



# Inversion prediction

```{r echo=FALSE}
#Load reference data scoreInvHap
data("Refs")
ref <- Refs$inv7p11.2
data("SNPsR2")
R2s <- SNPsR2$inv7p11.2
data("hetRefs")
hRefs <- hetRefs$inv7p11.2


byregion_inv <- scoreInvHap(SNPlist = byregion, SNPsR2 = R2s, hetRefs = hRefs, Refs = ref, imputed = TRUE, verbose = TRUE, BPPARAM = MulticoreParam(10))

whole_inv <- scoreInvHap(SNPlist = whole, SNPsR2 = R2s, hetRefs = hRefs, Refs = ref, imputed = TRUE, verbose = TRUE, BPPARAM = MulticoreParam(10))
```

Predicted inversions with scoreInvHap
```{r}
byregion_inv
whole_inv
```

```{r echo=FALSE}
scoreinvhap_table <- table(ByRegion = classification(byregion_inv)[sort(names(classification(byregion_inv)))], 
      WholeChr = classification(whole_inv)[sort(names(classification(whole_inv)))])
```

```{r}
# Comparison table
scoreinvhap_table
sum(diag(scoreinvhap_table))/sum(scoreinvhap_table)
kappa(scoreinvhap_table)
```


```{r}
# Comparison of the results for both imputation methods
par(mfrow=c(1,2))
hist(maxscores(byregion_inv), breaks=seq(0, 1, by=0.05), main="ByRegion", xlab="Maxscores")
hist(maxscores(whole_inv), breaks=seq(0, 1, by=0.05), main="WholeChr", xlab="Maxscores")
```

```{r echo=FALSE}
# Calculate scores correlation by individuals
score_corr <- mapply(cor, as.data.frame(scores(byregion_inv)[sort(rownames(scores(byregion_inv))),]),
                          as.data.frame(scores(whole_inv)[sort(rownames(scores(whole_inv))),]))
```

```{r}
# Score correlation by individuals between both imputation methods
min(score_corr)
max(score_corr)
mean(score_corr)
par(mfrow=c(1,1))
SCORE_CORR_HIST <- hist(score_corr, breaks=seq(0.9, 1, by=0.002), 
                        main="Score correlation by individuals", 
                        xlab="Score correlation")
```

```{r}
# Difference score between the highest similarity score and the second highest, 
# in both imputation methods
par(mfrow=c(1,2))
hist(diffscores(byregion_inv), breaks=seq(0, 1, by=0.05), 
     main="ByRegioin", xlab="Difference score")
hist(diffscores(whole_inv), breaks=seq(0, 1, by=0.05), 
     main="WholeChr", xlab="Difference score")
```


```{r}
# Numbers of scores used
mean(numSNPs(byregion_inv))
mean(numSNPs(whole_inv))
```

```{r}
# Number of samples in both imputation methods before and after QC filtering
length(classification(byregion_inv))
length(classification(byregion_inv, minDiff = 0.1, callRate = 0.9))
length(classification(byregion_inv, minDiff = 0.1, callRate = 0.9))/
  length(classification(byregion_inv))
length(classification(whole_inv))
length(classification(whole_inv, minDiff = 0.1, callRate = 0.9))
length(classification(whole_inv, minDiff = 0.1, callRate = 0.9))/
  length(classification(whole_inv))
```



```{r}
# Steps to compare the byregion with the nofilter_minimac
whole <- nofilter_minimac_7

whole <- whole[,colnames(byregion)]

# Run the code to compare imputation

noinv17 <- classification(nofilter_minimac_inv_7)[names(classification(byregion_inv_7))]

table(ByRegion = classification(byregion_inv_7)[sort(names(classification(byregion_inv_7)))],
       WholeChr = noinv17[sort(names(noinv17))])
# And run rest of the code to compare inversion prediction

```



